internal/controller/dnsrecord_groups.go

```diff
@@ -0,0 +1,327 @@
+					group := types.Group(g)
+					if len(g) > 0 && !activeGroups.HasGroup(group) {
+						activeGroups = append(activeGroups, group)
+					}
+				}
+			}
+
+		}
+	}
+
+	logger.V(1).Info("got active groups", "groups", activeGroups)
+	// no answers, return empty
+	return activeGroups
+}
+
+// getNameserversFromProvider extracts nameservers from the provider secret
+func (r *BaseDNSRecordReconciler) getNameserversFromProvider(ctx context.Context, c client.Client, dnsRecord DNSRecordAccessor) ([]string, error) {
+	var nameservers []string
+	var providerSecret v1.Secret
+	providerRef := dnsRecord.GetProviderRef()
+
+	if providerRef.Name != "" {
+		secretKey := client.ObjectKey{
+			Name:      providerRef.Name,
+			Namespace: dnsRecord.GetNamespace(),
+		}
+
+		if err := c.Get(ctx, secretKey, &providerSecret); err == nil {
+			nameservers = r.extractNameserversFromSecret(&providerSecret)
+		} else {
+			return nameservers, err
+		}
+	} else {
+		secretList := &v1.SecretList{}
+		err := c.List(ctx, secretList, &client.ListOptions{
+			LabelSelector: labels.SelectorFromSet(map[string]string{
+				v1alpha1.DefaultProviderSecretLabel: "true",
+			}),
+			Namespace: dnsRecord.GetNamespace(),
+		})
+
+		if err == nil && len(secretList.Items) == 1 {
+			nameservers = r.extractNameserversFromSecret(&secretList.Items[0])
+		}
+	}
+	return nameservers, nil
+}
+
+// extractNameserversFromSecret extracts and parses the NAMESERVERS field from a secret
+func (r *BaseDNSRecordReconciler) extractNameserversFromSecret(secret *v1.Secret) []string {
+	var nameservers []string
+	if secret == nil {
+		return nameservers
+	}
+	if nameserversData, ok := secret.Data["NAMESERVERS"]; ok && len(nameserversData) > 0 {
+		nameserversStr := string(nameserversData)
+		for ns := range strings.SplitSeq(nameserversStr, ",") {
+			ns = strings.TrimSpace(ns)
+			if ns != "" {
+				nameservers = append(nameservers, ns)
+			}
+		}
+	}
+
+	return nameservers
+}
+
+// unpublishInactiveGroups: handle unpublishing inactive groups records from the zone
+func (r *BaseDNSRecordReconciler) unpublishInactiveGroups(ctx context.Context, c client.Client, dnsRecord DNSRecordAccessor, dnsProvider provider.Provider) error {
+	logger := log.FromContext(ctx).WithName("active-groups").WithName("unpublish")
+	managedDNSRecordTypes := []string{externaldnsendpoint.RecordTypeA, externaldnsendpoint.RecordTypeAAAA, externaldnsendpoint.RecordTypeCNAME}
+
+	// if this record does not have a group, do not process unpublish
+	if dnsRecord.GetGroup() == "" {
+		return nil
+	}
+
+	activeGroups := r.getActiveGroups(ctx, c, dnsRecord)
+
+	// only process unpublish when there are active groups and we are reconciling a record from an active group
+	if len(activeGroups) == 0 || !dnsRecord.IsActive() {
+		return nil
+	}
+
+	if prematurely, _ := recordReceivedPrematurely(dnsRecord); prematurely {
+		logger.V(1).Info("Skipping DNSRecord - is still valid")
+		return nil
+	}
+
```

Comment:
Can this check ever be true? I am looking at how the `recordReceivedPrematurely` is handling the is active.

https://github.com/philbrookes/dns-operator/blob/gh-620/internal/controller/dnsrecord_controller.go#L584-L587

Reply:
good point, this does seem redundant now. I'll remove it